<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Scott Chamberlain" />

<meta name="date" content="2020-02-10" />

<title>Introduction to taxize</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to taxize</h1>
<h4 class="author">Scott Chamberlain</h4>
<h4 class="date">2020-02-10</h4>



<p><code>taxize</code> is a taxonomic toolbelt for R. <code>taxize</code> wraps APIs for a large suite of taxonomic databases availab on the web.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>First, install and load <code>taxize</code> into the R session.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">install.packages</span>(<span class="st">&quot;taxize&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">library</span>(<span class="st">&quot;taxize&quot;</span>)</span></code></pre></div>
<p>Advanced users can also download and install the latest development copy from <a href="https://github.com/ropensci/taxize_">GitHub</a>.</p>
</div>
<div id="resolve-taxonomic-name" class="section level2">
<h2>Resolve taxonomic name</h2>
<p>This is a common task in biology. We often have a list of species names and we want to know a) if we have the most up to date names, b) if our names are spelled correctly, and c) the scientific name for a common name. One way to resolve names is via the Global Names Resolver (GNR) service provided by the <a href="https://www.eol.org/">Encyclopedia of Life</a>. Here, we are searching for two misspelled names:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>temp &lt;-<span class="st"> </span><span class="kw">gnr_resolve</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;Helianthos annus&quot;</span>, <span class="st">&quot;Homo saapiens&quot;</span>))</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">head</span>(temp)</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; # A tibble: 6 x 5</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt;   user_supplied_name submitted_name  matched_name     data_source_title    score</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;           &lt;chr&gt;            &lt;chr&gt;                &lt;dbl&gt;</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; 1 Helianthos annus   Helianthos ann… Helianthus annus uBio NameBank         0.75</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt; 2 Helianthos annus   Helianthos ann… Helianthus annu… Catalogue of Life     0.75</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; 3 Helianthos annus   Helianthos ann… Helianthus annu… ITIS                  0.75</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; 4 Helianthos annus   Helianthos ann… Helianthus annu… NCBI                  0.75</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt; 5 Helianthos annus   Helianthos ann… Helianthus annu… GRIN Taxonomy for P…  0.75</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt; 6 Helianthos annus   Helianthos ann… Helianthus annu… Union 4               0.75</span></span></code></pre></div>
<p>The correct spellings are <em>Helianthus annuus</em> and <em>Homo sapiens</em>. Another approach uses the Taxonomic Name Resolution Service via the Taxosaurus API developed by iPLant and the Phylotastic organization. In this example, we provide a list of species names, some of which are misspelled, and we’ll call the API with the <em>tnrs</em> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>mynames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Helianthus annuus&quot;</span>, <span class="st">&quot;Pinus contort&quot;</span>, <span class="st">&quot;Poa anua&quot;</span>, <span class="st">&quot;Abis magnifica&quot;</span>,</span>
<span id="cb4-2"><a href="#cb4-2"></a>    <span class="st">&quot;Rosa california&quot;</span>, <span class="st">&quot;Festuca arundinace&quot;</span>, <span class="st">&quot;Sorbus occidentalos&quot;</span>,<span class="st">&quot;Madia sateva&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="kw">tnrs</span>(<span class="dt">query =</span> mynames, <span class="dt">source =</span> <span class="st">&quot;iPlant_TNRS&quot;</span>)[ , <span class="op">-</span><span class="kw">c</span>(<span class="dv">5</span><span class="op">:</span><span class="dv">7</span>)]</span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt;         submittedname        acceptedname    sourceid score</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt; 1   Helianthus annuus   Helianthus annuus iPlant_TNRS  1.00</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">#&gt; 2       Pinus contort      Pinus contorta iPlant_TNRS  0.98</span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">#&gt; 3            Poa anua           Poa annua iPlant_TNRS  0.96</span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">#&gt; 4      Abis magnifica     Abies magnifica iPlant_TNRS  0.96</span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">#&gt; 5     Rosa california    Rosa californica iPlant_TNRS  0.99</span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">#&gt; 6  Festuca arundinace Festuca arundinacea iPlant_TNRS  0.99</span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">#&gt; 7 Sorbus occidentalos Sorbus occidentalis iPlant_TNRS  0.99</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">#&gt; 8        Madia sateva        Madia sativa iPlant_TNRS  0.97</span></span></code></pre></div>
<p>It turns out there are a few corrections: e.g., <em>Madia sateva</em> should be <em>Madia sativa</em>, and <em>Rosa california</em> should be <em>Rosa californica</em>. Note that this search worked because fuzzy matching was employed to retrieve names that were close, but not exact matches. Fuzzy matching is only available for plants in the TNRS service, so we advise using EOL’s Global Names Resolver if you need to resolve animal names.</p>
<p>taxize takes the approach that the user should be able to make decisions about what resource to trust, rather than making the decision. Both the EOL GNR and the TNRS services provide data from a variety of data sources. The user may trust a specific data source, thus may want to use the names from that data source. In the future, we may provide the ability for taxize to suggest the best match from a variety of sources.</p>
<p>Another common use case is when there are many synonyms for a species. In this example, we have three synonyms of the currently accepted name for a species.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>mynames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Helianthus annuus ssp. jaegeri&quot;</span>, <span class="st">&quot;Helianthus annuus ssp. lenticularis&quot;</span>, <span class="st">&quot;Helianthus annuus ssp. texanus&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2"></a>(tsn &lt;-<span class="st"> </span><span class="kw">get_tsn</span>(mynames, <span class="dt">accepted =</span> <span class="ot">FALSE</span>))</span>
<span id="cb5-3"><a href="#cb5-3"></a>══  <span class="dv">3</span> queries  ═══════════════</span>
<span id="cb5-4"><a href="#cb5-4"></a>✔  Found<span class="op">:</span><span class="st">  </span>Helianthus annuus ssp. jaegeri</span>
<span id="cb5-5"><a href="#cb5-5"></a>✔  Found<span class="op">:</span><span class="st">  </span>Helianthus annuus ssp. lenticularis</span>
<span id="cb5-6"><a href="#cb5-6"></a>✔  Found<span class="op">:</span><span class="st">  </span>Helianthus annuus ssp. texanus</span>
<span id="cb5-7"><a href="#cb5-7"></a>══  Results  ═════════════════</span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a>● Total<span class="op">:</span><span class="st"> </span><span class="dv">3</span> </span>
<span id="cb5-10"><a href="#cb5-10"></a>● Found<span class="op">:</span><span class="st"> </span><span class="dv">3</span> </span>
<span id="cb5-11"><a href="#cb5-11"></a>● Not Found<span class="op">:</span><span class="st"> </span><span class="dv">0</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>[<span class="dv">1</span>] <span class="st">&quot;525928&quot;</span> <span class="st">&quot;525929&quot;</span> <span class="st">&quot;525930&quot;</span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="kw">attr</span>(,<span class="st">&quot;class&quot;</span>)</span>
<span id="cb5-14"><a href="#cb5-14"></a>[<span class="dv">1</span>] <span class="st">&quot;tsn&quot;</span></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="kw">attr</span>(,<span class="st">&quot;match&quot;</span>)</span>
<span id="cb5-16"><a href="#cb5-16"></a>[<span class="dv">1</span>] <span class="st">&quot;found&quot;</span> <span class="st">&quot;found&quot;</span> <span class="st">&quot;found&quot;</span></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="kw">attr</span>(,<span class="st">&quot;multiple_matches&quot;</span>)</span>
<span id="cb5-18"><a href="#cb5-18"></a>[<span class="dv">1</span>] <span class="ot">FALSE</span> <span class="ot">FALSE</span> <span class="ot">FALSE</span></span>
<span id="cb5-19"><a href="#cb5-19"></a><span class="kw">attr</span>(,<span class="st">&quot;pattern_match&quot;</span>)</span>
<span id="cb5-20"><a href="#cb5-20"></a>[<span class="dv">1</span>] <span class="ot">FALSE</span> <span class="ot">FALSE</span> <span class="ot">FALSE</span></span>
<span id="cb5-21"><a href="#cb5-21"></a><span class="kw">attr</span>(,<span class="st">&quot;uri&quot;</span>)</span>
<span id="cb5-22"><a href="#cb5-22"></a>[<span class="dv">1</span>] <span class="st">&quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=525928&quot;</span></span>
<span id="cb5-23"><a href="#cb5-23"></a>[<span class="dv">2</span>] <span class="st">&quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=525929&quot;</span></span>
<span id="cb5-24"><a href="#cb5-24"></a>[<span class="dv">3</span>] <span class="st">&quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=525930&quot;</span></span>
<span id="cb5-25"><a href="#cb5-25"></a><span class="kw">lapply</span>(tsn, itis_acceptname)</span>
<span id="cb5-26"><a href="#cb5-26"></a>[[<span class="dv">1</span>]]</span>
<span id="cb5-27"><a href="#cb5-27"></a>  submittedtsn      acceptedname acceptedtsn author</span>
<span id="cb5-28"><a href="#cb5-28"></a><span class="dv">1</span>       <span class="dv">525928</span> Helianthus annuus       <span class="dv">36616</span>     L.</span>
<span id="cb5-29"><a href="#cb5-29"></a></span>
<span id="cb5-30"><a href="#cb5-30"></a>[[<span class="dv">2</span>]]</span>
<span id="cb5-31"><a href="#cb5-31"></a>  submittedtsn      acceptedname acceptedtsn author</span>
<span id="cb5-32"><a href="#cb5-32"></a><span class="dv">1</span>       <span class="dv">525929</span> Helianthus annuus       <span class="dv">36616</span>     L.</span>
<span id="cb5-33"><a href="#cb5-33"></a></span>
<span id="cb5-34"><a href="#cb5-34"></a>[[<span class="dv">3</span>]]</span>
<span id="cb5-35"><a href="#cb5-35"></a>  submittedtsn      acceptedname acceptedtsn author</span>
<span id="cb5-36"><a href="#cb5-36"></a><span class="dv">1</span>       <span class="dv">525930</span> Helianthus annuus       <span class="dv">36616</span>     L.</span></code></pre></div>
</div>
<div id="retrieve-higher-taxonomic-names" class="section level2">
<h2>Retrieve higher taxonomic names</h2>
<p>Another task biologists often face is getting higher taxonomic names for a taxa list. Having the higher taxonomy allows you to put into context the relationships of your species list. For example, you may find out that species A and species B are in Family C, which may lead to some interesting insight, as opposed to not knowing that Species A and B are closely related. This also makes it easy to aggregate/standardize data to a specific taxonomic level (e.g., family level) or to match data to other databases with different taxonomic resolution (e.g., trait databases).</p>
<p>A number of data sources in taxize provide the capability to retrieve higher taxonomic names, but we will highlight two of the more useful ones: <a href="https://www.itis.gov/">Integrated Taxonomic Information System (ITIS)</a> and <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (NCBI)</a>. First, we’ll search for two species, <em>Abies procera} and </em>Pinus contorta* within ITIS.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>specieslist &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Abies procera&quot;</span>,<span class="st">&quot;Pinus contorta&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">classification</span>(specieslist, <span class="dt">db =</span> <span class="st">&#39;itis&#39;</span>)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt; ══  2 queries  ═══════════════</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; ✔  Found:  Abies procera</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; ✔  Found:  Pinus contorta</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">#&gt; </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">#&gt; ● Total: 2 </span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">#&gt; ● Found: 2 </span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">#&gt; $`Abies procera`</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">#&gt;               name          rank     id</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">#&gt; 1          Plantae       kingdom 202422</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">#&gt; 2    Viridiplantae    subkingdom 954898</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co">#&gt; 3     Streptophyta  infrakingdom 846494</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co">#&gt; 4      Embryophyta superdivision 954900</span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="co">#&gt; 5     Tracheophyta      division 846496</span></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co">#&gt; 6  Spermatophytina   subdivision 846504</span></span>
<span id="cb6-19"><a href="#cb6-19"></a><span class="co">#&gt; 7        Pinopsida         class 500009</span></span>
<span id="cb6-20"><a href="#cb6-20"></a><span class="co">#&gt; 8          Pinidae      subclass 954916</span></span>
<span id="cb6-21"><a href="#cb6-21"></a><span class="co">#&gt; 9          Pinales         order 500028</span></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="co">#&gt; 10        Pinaceae        family  18030</span></span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="co">#&gt; 11           Abies         genus  18031</span></span>
<span id="cb6-24"><a href="#cb6-24"></a><span class="co">#&gt; 12   Abies procera       species 181835</span></span>
<span id="cb6-25"><a href="#cb6-25"></a><span class="co">#&gt; </span></span>
<span id="cb6-26"><a href="#cb6-26"></a><span class="co">#&gt; $`Pinus contorta`</span></span>
<span id="cb6-27"><a href="#cb6-27"></a><span class="co">#&gt;               name          rank     id</span></span>
<span id="cb6-28"><a href="#cb6-28"></a><span class="co">#&gt; 1          Plantae       kingdom 202422</span></span>
<span id="cb6-29"><a href="#cb6-29"></a><span class="co">#&gt; 2    Viridiplantae    subkingdom 954898</span></span>
<span id="cb6-30"><a href="#cb6-30"></a><span class="co">#&gt; 3     Streptophyta  infrakingdom 846494</span></span>
<span id="cb6-31"><a href="#cb6-31"></a><span class="co">#&gt; 4      Embryophyta superdivision 954900</span></span>
<span id="cb6-32"><a href="#cb6-32"></a><span class="co">#&gt; 5     Tracheophyta      division 846496</span></span>
<span id="cb6-33"><a href="#cb6-33"></a><span class="co">#&gt; 6  Spermatophytina   subdivision 846504</span></span>
<span id="cb6-34"><a href="#cb6-34"></a><span class="co">#&gt; 7        Pinopsida         class 500009</span></span>
<span id="cb6-35"><a href="#cb6-35"></a><span class="co">#&gt; 8          Pinidae      subclass 954916</span></span>
<span id="cb6-36"><a href="#cb6-36"></a><span class="co">#&gt; 9          Pinales         order 500028</span></span>
<span id="cb6-37"><a href="#cb6-37"></a><span class="co">#&gt; 10        Pinaceae        family  18030</span></span>
<span id="cb6-38"><a href="#cb6-38"></a><span class="co">#&gt; 11           Pinus         genus  18035</span></span>
<span id="cb6-39"><a href="#cb6-39"></a><span class="co">#&gt; 12  Pinus contorta       species 183327</span></span>
<span id="cb6-40"><a href="#cb6-40"></a><span class="co">#&gt; </span></span>
<span id="cb6-41"><a href="#cb6-41"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb6-42"><a href="#cb6-42"></a><span class="co">#&gt; [1] &quot;classification&quot;</span></span>
<span id="cb6-43"><a href="#cb6-43"></a><span class="co">#&gt; attr(,&quot;db&quot;)</span></span>
<span id="cb6-44"><a href="#cb6-44"></a><span class="co">#&gt; [1] &quot;itis&quot;</span></span></code></pre></div>
<p>It turns out both species are in the family Pinaceae. You can also get this type of information from the NCBI by doing <code>classification(specieslist, db = &#39;ncbi&#39;)</code>.</p>
<p>Instead of a full classification, you may only want a single name, say a family name for your species of interest. The function <code>tax_name</code> is built just for this purpose. As with the <code>classification</code> function you can specify the data source with the <code>db</code> argument, either ITIS or NCBI.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">tax_name</span>(<span class="dt">query =</span> <span class="st">&quot;Helianthus annuus&quot;</span>, <span class="dt">get =</span> <span class="st">&quot;family&quot;</span>, <span class="dt">db =</span> <span class="st">&quot;ncbi&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#&gt; ✔  Found:  Helianthus+annuus</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt; </span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt; ● Found: 1 </span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">#&gt;     db             query     family</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">#&gt; 1 ncbi Helianthus annuus Asteraceae</span></span></code></pre></div>
<p>It may happen that a data source does not provide information on the queried species, than one could take the result from another source and union the results from the different sources.</p>
</div>
<div id="interactive-name-selection" class="section level2">
<h2>Interactive name selection</h2>
<p>As mentioned most databases use a numeric code to reference a species. A general workflow in taxize is: Retrieve Code for the queried species and then use this code to query more data/information.</p>
<p>Below are a few examples. When you run these examples in R, you are presented with a command prompt asking for the row that contains the name you would like back; that output is not printed below for brevity. In this example, the search term has many matches. The function returns a data frame of the matches, and asks for the user to input what row number to accept.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">get_uid</span>(<span class="dt">sciname =</span> <span class="st">&quot;Pinus&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">#&gt;   status     rank    division scientificname commonname    uid genus species</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; 1 active subgenus seed plants          Pinus hard pines 139271              </span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#&gt; 2 active    genus seed plants          Pinus              3337              </span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">#&gt;   subsp modificationdate</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; 1       2017/06/14 00:00</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#&gt; 2       2016/03/25 00:00</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">#&gt; </span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">#&gt; ● Found: 0 </span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co">#&gt; [1] NA</span></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="co">#&gt; [1] &quot;uid&quot;</span></span>
<span id="cb8-17"><a href="#cb8-17"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb8-18"><a href="#cb8-18"></a><span class="co">#&gt; [1] &quot;NA due to user input out of range&quot;</span></span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">#&gt; [1] TRUE</span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>In another example, you can pass in a long character vector of taxonomic names (although this one is rather short for demo purposes):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>splist &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;annona cherimola&quot;</span>, <span class="st">&#39;annona muricata&#39;</span>, <span class="st">&quot;quercus robur&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="kw">get_tsn</span>(<span class="dt">searchterm =</span> splist, <span class="dt">searchtype =</span> <span class="st">&quot;scientific&quot;</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">#&gt; ══  3 queries  ═══════════════</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co">#&gt; ✔  Found:  annona cherimola</span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">#&gt; ✔  Found:  annona muricata</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co">#&gt; ✔  Found:  quercus robur</span></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">#&gt; </span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">#&gt; ● Total: 3 </span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co">#&gt; ● Found: 3 </span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co">#&gt; [1] &quot;506198&quot; &quot;18098&quot;  &quot;19405&quot; </span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="co">#&gt; [1] &quot;tsn&quot;</span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="co">#&gt; [1] &quot;found&quot; &quot;found&quot; &quot;found&quot;</span></span>
<span id="cb9-17"><a href="#cb9-17"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb9-18"><a href="#cb9-18"></a><span class="co">#&gt; [1] FALSE FALSE  TRUE</span></span>
<span id="cb9-19"><a href="#cb9-19"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb9-20"><a href="#cb9-20"></a><span class="co">#&gt; [1] FALSE FALSE  TRUE</span></span>
<span id="cb9-21"><a href="#cb9-21"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb9-22"><a href="#cb9-22"></a><span class="co">#&gt; [1] &quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=506198&quot;</span></span>
<span id="cb9-23"><a href="#cb9-23"></a><span class="co">#&gt; [2] &quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=18098&quot; </span></span>
<span id="cb9-24"><a href="#cb9-24"></a><span class="co">#&gt; [3] &quot;https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&amp;search_value=19405&quot;</span></span></code></pre></div>
<p>There are functions for many other sources</p>
<ul>
<li><code>get_boldid()</code></li>
<li><code>get_eolid()</code></li>
<li><code>get_gbifid()</code></li>
<li><code>get_nbnid()</code></li>
<li><code>get_tpsid()</code></li>
</ul>
<p>Sometimes with these functions you get a lot of data back. In these cases you may want to limit your choices. Soon we will incorporate the ability to filter using <code>regex</code> to limit matches, but for now, we have a new parameter, <code>rows</code>, which lets you select certain rows. For example, you can select the first row of each given name, which means there is no interactive component:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">get_nbnid</span>(<span class="kw">c</span>(<span class="st">&quot;Zootoca vivipara&quot;</span>,<span class="st">&quot;Pinus contorta&quot;</span>), <span class="dt">rows =</span> <span class="dv">1</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co">#&gt; ══  2 queries  ═══════════════</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt; ✔  Found:  Zootoca vivipara</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt; ✔  Found:  Pinus contorta</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt; </span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt; ● Total: 2 </span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt; ● Found: 2 </span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt; [1] &quot;NHMSYS0001706186&quot; &quot;NBNSYS0000004786&quot;</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="co">#&gt; [1] &quot;nbnid&quot;</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="co">#&gt; [1] &quot;found&quot; &quot;found&quot;</span></span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb10-16"><a href="#cb10-16"></a><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb10-18"><a href="#cb10-18"></a><span class="co">#&gt; [1] FALSE FALSE</span></span>
<span id="cb10-19"><a href="#cb10-19"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb10-20"><a href="#cb10-20"></a><span class="co">#&gt; [1] &quot;https://species.nbnatlas.org/species/NHMSYS0001706186&quot;</span></span>
<span id="cb10-21"><a href="#cb10-21"></a><span class="co">#&gt; [2] &quot;https://species.nbnatlas.org/species/NBNSYS0000004786&quot;</span></span></code></pre></div>
<p>Or you can select a range of rows</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">get_nbnid</span>(<span class="kw">c</span>(<span class="st">&quot;Zootoca vivipara&quot;</span>,<span class="st">&quot;Pinus contorta&quot;</span>), <span class="dt">rows =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co">#&gt; ══  2 queries  ═══════════════</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co">#&gt; ✔  Found:  Zootoca vivipara</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co">#&gt; ✔  Found:  Pinus contorta</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co">#&gt; </span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co">#&gt; ● Total: 2 </span></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co">#&gt; ● Found: 2 </span></span>
<span id="cb11-9"><a href="#cb11-9"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="co">#&gt; [1] &quot;NHMSYS0001706186&quot; &quot;NBNSYS0000004786&quot;</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co">#&gt; [1] &quot;nbnid&quot;</span></span>
<span id="cb11-13"><a href="#cb11-13"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb11-14"><a href="#cb11-14"></a><span class="co">#&gt; [1] &quot;found&quot; &quot;found&quot;</span></span>
<span id="cb11-15"><a href="#cb11-15"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb11-16"><a href="#cb11-16"></a><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span id="cb11-17"><a href="#cb11-17"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb11-18"><a href="#cb11-18"></a><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span id="cb11-19"><a href="#cb11-19"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb11-20"><a href="#cb11-20"></a><span class="co">#&gt; [1] &quot;https://species.nbnatlas.org/species/NHMSYS0001706186&quot;</span></span>
<span id="cb11-21"><a href="#cb11-21"></a><span class="co">#&gt; [2] &quot;https://species.nbnatlas.org/species/NBNSYS0000004786&quot;</span></span></code></pre></div>
<p>In addition, in case you don’t want to do interactive name selection in the case where there are a lot of names, you can get all data back with functions of the form, e.g., <code>get_tsn_()</code>, and likewise for other data sources. For example:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>out &lt;-<span class="st"> </span><span class="kw">get_nbnid_</span>(<span class="st">&quot;Poa annua&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="kw">NROW</span>(out<span class="op">$</span><span class="st">`</span><span class="dt">Poa annua</span><span class="st">`</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">#&gt; [1] 25</span></span></code></pre></div>
<p>That’s a lot of data, so we can get only certain rows back</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">get_nbnid_</span>(<span class="st">&quot;Poa annua&quot;</span>, <span class="dt">rows =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co">#&gt; $`Poa annua`</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="co">#&gt;                guid     scientificName    rank taxonomicStatus</span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co">#&gt; 1  NBNSYS0000002544          Poa annua species        accepted</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co">#&gt; 2  NBNSYS0200001901       Bellis annua species        accepted</span></span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="co">#&gt; 3  NBNSYS0200003392   Triumfetta annua species        accepted</span></span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="co">#&gt; 4  NBNSYS0200002555        Lonas annua species        accepted</span></span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="co">#&gt; 5  NHMSYS0000456951  Carrichtera annua species        accepted</span></span>
<span id="cb13-9"><a href="#cb13-9"></a><span class="co">#&gt; 6  NHMSYS0000461817     Poa sieberiana species        accepted</span></span>
<span id="cb13-10"><a href="#cb13-10"></a><span class="co">#&gt; 7  NHMSYS0000461808      Poa ligularis species        accepted</span></span>
<span id="cb13-11"><a href="#cb13-11"></a><span class="co">#&gt; 8  NHMSYS0000461807 Poa labillardierei species        accepted</span></span>
<span id="cb13-12"><a href="#cb13-12"></a><span class="co">#&gt; 9  NHMSYS0000461805         Poa gunnii species        accepted</span></span>
<span id="cb13-13"><a href="#cb13-13"></a><span class="co">#&gt; 10 NHMSYS0000461801     Poa costiniana species        accepted</span></span></code></pre></div>
</div>
<div id="coerce-numericsalphanumerics-to-taxon-ids" class="section level2">
<h2>Coerce numerics/alphanumerics to taxon IDs</h2>
<p>We’ve also introduced in <code>v0.5</code> the ability to coerce numerics and alphanumerics to taxonomic ID classes that are usually only retrieved via <code>get_*()</code> functions.</p>
<p>For example, adfafd</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">as.gbifid</span>(<span class="kw">get_gbifid</span>(<span class="st">&quot;Poa annua&quot;</span>)) <span class="co"># already a uid, returns the same</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co">#&gt;    gbifid             scientificname    rank   status matchtype</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">#&gt; 1 2704179               Poa annua L. species ACCEPTED     EXACT</span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">#&gt; 2 8422205 Poa annua Cham. &amp; Schltdl. species  SYNONYM     EXACT</span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co">#&gt; 3 7730008           Poa annua Steud. species DOUBTFUL     EXACT</span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co">#&gt; ✖  Not Found:  Poa annua</span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">#&gt; </span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">#&gt; ● Found: 0 </span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">#&gt; ● Not Found: 1</span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co">#&gt; [1] NA</span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="co">#&gt; [1] &quot;gbifid&quot;</span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb14-17"><a href="#cb14-17"></a><span class="co">#&gt; [1] &quot;not found&quot;</span></span>
<span id="cb14-18"><a href="#cb14-18"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb14-19"><a href="#cb14-19"></a><span class="co">#&gt; [1] TRUE</span></span>
<span id="cb14-20"><a href="#cb14-20"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb14-21"><a href="#cb14-21"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb14-22"><a href="#cb14-22"></a><span class="kw">as.gbifid</span>(<span class="dv">2704179</span>) <span class="co"># numeric</span></span>
<span id="cb14-23"><a href="#cb14-23"></a><span class="co">#&gt; [1] &quot;2704179&quot;</span></span>
<span id="cb14-24"><a href="#cb14-24"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb14-25"><a href="#cb14-25"></a><span class="co">#&gt; [1] &quot;gbifid&quot;</span></span>
<span id="cb14-26"><a href="#cb14-26"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb14-27"><a href="#cb14-27"></a><span class="co">#&gt; [1] &quot;found&quot;</span></span>
<span id="cb14-28"><a href="#cb14-28"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb14-29"><a href="#cb14-29"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb14-30"><a href="#cb14-30"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb14-31"><a href="#cb14-31"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb14-32"><a href="#cb14-32"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb14-33"><a href="#cb14-33"></a><span class="co">#&gt; [1] &quot;http://www.gbif.org/species/2704179&quot;</span></span>
<span id="cb14-34"><a href="#cb14-34"></a><span class="kw">as.gbifid</span>(<span class="st">&quot;2704179&quot;</span>) <span class="co"># character</span></span>
<span id="cb14-35"><a href="#cb14-35"></a><span class="co">#&gt; [1] &quot;2704179&quot;</span></span>
<span id="cb14-36"><a href="#cb14-36"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb14-37"><a href="#cb14-37"></a><span class="co">#&gt; [1] &quot;gbifid&quot;</span></span>
<span id="cb14-38"><a href="#cb14-38"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb14-39"><a href="#cb14-39"></a><span class="co">#&gt; [1] &quot;found&quot;</span></span>
<span id="cb14-40"><a href="#cb14-40"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb14-41"><a href="#cb14-41"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb14-42"><a href="#cb14-42"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb14-43"><a href="#cb14-43"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb14-44"><a href="#cb14-44"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb14-45"><a href="#cb14-45"></a><span class="co">#&gt; [1] &quot;http://www.gbif.org/species/2704179&quot;</span></span>
<span id="cb14-46"><a href="#cb14-46"></a><span class="kw">as.gbifid</span>(<span class="kw">list</span>(<span class="st">&quot;2704179&quot;</span>,<span class="st">&quot;2435099&quot;</span>,<span class="st">&quot;3171445&quot;</span>)) <span class="co"># list, either numeric or character</span></span>
<span id="cb14-47"><a href="#cb14-47"></a><span class="co">#&gt; [1] &quot;2704179&quot; &quot;2435099&quot; &quot;3171445&quot;</span></span>
<span id="cb14-48"><a href="#cb14-48"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb14-49"><a href="#cb14-49"></a><span class="co">#&gt; [1] &quot;gbifid&quot;</span></span>
<span id="cb14-50"><a href="#cb14-50"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb14-51"><a href="#cb14-51"></a><span class="co">#&gt; [1] &quot;found&quot; &quot;found&quot; &quot;found&quot;</span></span>
<span id="cb14-52"><a href="#cb14-52"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb14-53"><a href="#cb14-53"></a><span class="co">#&gt; [1] FALSE FALSE FALSE</span></span>
<span id="cb14-54"><a href="#cb14-54"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb14-55"><a href="#cb14-55"></a><span class="co">#&gt; [1] FALSE FALSE FALSE</span></span>
<span id="cb14-56"><a href="#cb14-56"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb14-57"><a href="#cb14-57"></a><span class="co">#&gt; [1] &quot;http://www.gbif.org/species/2704179&quot; &quot;http://www.gbif.org/species/2435099&quot;</span></span>
<span id="cb14-58"><a href="#cb14-58"></a><span class="co">#&gt; [3] &quot;http://www.gbif.org/species/3171445&quot;</span></span></code></pre></div>
<p>These <code>as.*()</code> functions do a quick check of the web resource to make sure it’s a real ID. However, you can turn this check off, making this coercion much faster:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">system.time</span>( <span class="kw">replicate</span>(<span class="dv">3</span>, <span class="kw">as.gbifid</span>(<span class="kw">c</span>(<span class="st">&quot;2704179&quot;</span>,<span class="st">&quot;2435099&quot;</span>,<span class="st">&quot;3171445&quot;</span>), <span class="dt">check=</span><span class="ot">TRUE</span>)) )</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="co">#&gt;   0.077   0.002   4.616</span></span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="kw">system.time</span>( <span class="kw">replicate</span>(<span class="dv">3</span>, <span class="kw">as.gbifid</span>(<span class="kw">c</span>(<span class="st">&quot;2704179&quot;</span>,<span class="st">&quot;2435099&quot;</span>,<span class="st">&quot;3171445&quot;</span>), <span class="dt">check=</span><span class="ot">FALSE</span>)) )</span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="co">#&gt;   0.002   0.000   0.002</span></span></code></pre></div>
</div>
<div id="what-taxa-are-downstream-of-my-taxon-of-interest" class="section level2">
<h2>What taxa are downstream of my taxon of interest?</h2>
<p>If someone is not a taxonomic specialist on a particular taxon he likely does not know what children taxa are within a family, or within a genus. This task becomes especially unwieldy when there are a large number of taxa downstream. You can of course go to a website like <a href="https://species.wikimedia.org/wiki/Main_Page">Wikispecies</a> or <a href="https://www.eol.org/">Encyclopedia of Life</a> to get downstream names. However, taxize provides an easy way to programatically search for downstream taxa for the <a href="https://www.itis.gov/">Integrated Taxonomic Information System</a>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>apis_itis_id &lt;-<span class="st"> </span><span class="dv">154395</span> <span class="co"># id for Apis, fetched beforehand to save time here</span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="kw">downstream</span>(apis_itis_id, <span class="dt">downto =</span> <span class="st">&quot;species&quot;</span>, <span class="dt">db =</span> <span class="st">&quot;itis&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="co">#&gt; $`154395`</span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="co">#&gt;      tsn parentname parenttsn rankname          taxonname rankid</span></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co">#&gt; 1 154396       Apis    154395  species     Apis mellifera    220</span></span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="co">#&gt; 2 763550       Apis    154395  species Apis andreniformis    220</span></span>
<span id="cb16-7"><a href="#cb16-7"></a><span class="co">#&gt; 3 763551       Apis    154395  species        Apis cerana    220</span></span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="co">#&gt; 4 763552       Apis    154395  species       Apis dorsata    220</span></span>
<span id="cb16-9"><a href="#cb16-9"></a><span class="co">#&gt; 5 763553       Apis    154395  species        Apis florea    220</span></span>
<span id="cb16-10"><a href="#cb16-10"></a><span class="co">#&gt; 6 763554       Apis    154395  species Apis koschevnikovi    220</span></span>
<span id="cb16-11"><a href="#cb16-11"></a><span class="co">#&gt; 7 763555       Apis    154395  species   Apis nigrocincta    220</span></span>
<span id="cb16-12"><a href="#cb16-12"></a><span class="co">#&gt; </span></span>
<span id="cb16-13"><a href="#cb16-13"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb16-14"><a href="#cb16-14"></a><span class="co">#&gt; [1] &quot;downstream&quot;</span></span>
<span id="cb16-15"><a href="#cb16-15"></a><span class="co">#&gt; attr(,&quot;db&quot;)</span></span>
<span id="cb16-16"><a href="#cb16-16"></a><span class="co">#&gt; [1] &quot;itis&quot;</span></span></code></pre></div>
</div>
<div id="direct-children" class="section level2">
<h2>Direct children</h2>
<p>You may sometimes only want the direct children. We got you covered on that front, with methods for ITIS and NCBI.</p>
<p>The direct children (genera in this case) of <em>Pinaceae</em> using NCBI data:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="kw">children</span>(<span class="st">&quot;Pinaceae&quot;</span>, <span class="dt">db =</span> <span class="st">&quot;ncbi&quot;</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co">#&gt; $Pinaceae</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">#&gt;    childtaxa_id childtaxa_name childtaxa_rank</span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; 1        123600     Nothotsuga          genus</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt; 2         64685        Cathaya          genus</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt; 3          3358          Tsuga          genus</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; 4          3356    Pseudotsuga          genus</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt; 5          3354    Pseudolarix          genus</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">#&gt; 6          3337          Pinus          genus</span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">#&gt; 7          3328          Picea          genus</span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">#&gt; 8          3325          Larix          genus</span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">#&gt; 9          3323     Keteleeria          genus</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">#&gt; 10         3321         Cedrus          genus</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="co">#&gt; 11         3319          Abies          genus</span></span>
<span id="cb17-15"><a href="#cb17-15"></a><span class="co">#&gt; </span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="co">#&gt; [1] &quot;children&quot;</span></span>
<span id="cb17-18"><a href="#cb17-18"></a><span class="co">#&gt; attr(,&quot;db&quot;)</span></span>
<span id="cb17-19"><a href="#cb17-19"></a><span class="co">#&gt; [1] &quot;ncbi&quot;</span></span></code></pre></div>
</div>
<div id="get-ncbi-id-from-genbank-ids" class="section level2">
<h2>Get NCBI ID from GenBank Ids</h2>
<p>With accession numbers</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">genbank2uid</span>(<span class="dt">id =</span> <span class="st">&#39;AJ748748&#39;</span>)</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="co">#&gt; [1] &quot;282199&quot;</span></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">#&gt; [1] &quot;uid&quot;</span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">#&gt; [1] &quot;found&quot;</span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co">#&gt; [1] &quot;https://www.ncbi.nlm.nih.gov/taxonomy/282199&quot;</span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="co">#&gt; attr(,&quot;name&quot;)</span></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co">#&gt; [1] &quot;Nereida ignava 16S rRNA gene, type strain 2SM4T&quot;</span></span></code></pre></div>
<p>With gi numbers</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">genbank2uid</span>(<span class="dt">id =</span> <span class="dv">62689767</span>)</span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co">#&gt; [1] &quot;282199&quot;</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co">#&gt; [1] &quot;uid&quot;</span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">#&gt; attr(,&quot;match&quot;)</span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co">#&gt; [1] &quot;found&quot;</span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">#&gt; attr(,&quot;multiple_matches&quot;)</span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">#&gt; attr(,&quot;pattern_match&quot;)</span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="co">#&gt; attr(,&quot;uri&quot;)</span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">#&gt; [1] &quot;https://www.ncbi.nlm.nih.gov/taxonomy/282199&quot;</span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co">#&gt; attr(,&quot;name&quot;)</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="co">#&gt; [1] &quot;Nereida ignava 16S rRNA gene, type strain 2SM4T&quot;</span></span></code></pre></div>
</div>
<div id="matching-species-tables-with-different-taxonomic-resolution" class="section level2">
<h2>Matching species tables with different taxonomic resolution</h2>
<p>Biologist often need to match different sets of data tied to species. For example, trait-based approaches are a promising tool in ecology. One problem is that abundance data must be matched with trait databases. These two data tables may contain species information on different taxonomic levels and possibly data must be aggregated to a joint taxonomic level, so that the data can be merged. taxize can help in this data-cleaning step, providing a reproducible workflow:</p>
<p>We can use the mentioned <code>classification</code>-function to retrieve the taxonomic hierarchy and then search the hierarchies up- and downwards for matches. Here is an example to match a species with names on three different taxonomic levels.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>A &lt;-<span class="st"> &quot;gammarus roeseli&quot;</span></span>
<span id="cb20-2"><a href="#cb20-2"></a></span>
<span id="cb20-3"><a href="#cb20-3"></a>B1 &lt;-<span class="st"> &quot;gammarus roeseli&quot;</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>B2 &lt;-<span class="st"> &quot;gammarus&quot;</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>B3 &lt;-<span class="st"> &quot;gammaridae&quot;</span></span>
<span id="cb20-6"><a href="#cb20-6"></a></span>
<span id="cb20-7"><a href="#cb20-7"></a>A_clas &lt;-<span class="st"> </span><span class="kw">classification</span>(A, <span class="dt">db =</span> <span class="st">&#39;ncbi&#39;</span>)</span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="co">#&gt; ✔  Found:  gammarus+roeseli</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="co">#&gt; </span></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="co">#&gt; ● Found: 1 </span></span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb20-15"><a href="#cb20-15"></a>B1_clas &lt;-<span class="st"> </span><span class="kw">classification</span>(B1, <span class="dt">db =</span> <span class="st">&#39;ncbi&#39;</span>)</span>
<span id="cb20-16"><a href="#cb20-16"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb20-17"><a href="#cb20-17"></a><span class="co">#&gt; ✔  Found:  gammarus+roeseli</span></span>
<span id="cb20-18"><a href="#cb20-18"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb20-19"><a href="#cb20-19"></a><span class="co">#&gt; </span></span>
<span id="cb20-20"><a href="#cb20-20"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb20-21"><a href="#cb20-21"></a><span class="co">#&gt; ● Found: 1 </span></span>
<span id="cb20-22"><a href="#cb20-22"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb20-23"><a href="#cb20-23"></a>B2_clas &lt;-<span class="st"> </span><span class="kw">classification</span>(B2, <span class="dt">db =</span> <span class="st">&#39;ncbi&#39;</span>)</span>
<span id="cb20-24"><a href="#cb20-24"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb20-25"><a href="#cb20-25"></a><span class="co">#&gt; ✔  Found:  gammarus</span></span>
<span id="cb20-26"><a href="#cb20-26"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb20-27"><a href="#cb20-27"></a><span class="co">#&gt; </span></span>
<span id="cb20-28"><a href="#cb20-28"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb20-29"><a href="#cb20-29"></a><span class="co">#&gt; ● Found: 1 </span></span>
<span id="cb20-30"><a href="#cb20-30"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb20-31"><a href="#cb20-31"></a>B3_clas &lt;-<span class="st"> </span><span class="kw">classification</span>(B3, <span class="dt">db =</span> <span class="st">&#39;ncbi&#39;</span>)</span>
<span id="cb20-32"><a href="#cb20-32"></a><span class="co">#&gt; ══  1 queries  ═══════════════</span></span>
<span id="cb20-33"><a href="#cb20-33"></a><span class="co">#&gt; ✔  Found:  gammaridae</span></span>
<span id="cb20-34"><a href="#cb20-34"></a><span class="co">#&gt; ══  Results  ═════════════════</span></span>
<span id="cb20-35"><a href="#cb20-35"></a><span class="co">#&gt; </span></span>
<span id="cb20-36"><a href="#cb20-36"></a><span class="co">#&gt; ● Total: 1 </span></span>
<span id="cb20-37"><a href="#cb20-37"></a><span class="co">#&gt; ● Found: 1 </span></span>
<span id="cb20-38"><a href="#cb20-38"></a><span class="co">#&gt; ● Not Found: 0</span></span>
<span id="cb20-39"><a href="#cb20-39"></a></span>
<span id="cb20-40"><a href="#cb20-40"></a>B1[<span class="kw">match</span>(A, B1)]</span>
<span id="cb20-41"><a href="#cb20-41"></a><span class="co">#&gt; [1] &quot;gammarus roeseli&quot;</span></span>
<span id="cb20-42"><a href="#cb20-42"></a>A_clas[[<span class="dv">1</span>]]<span class="op">$</span>rank[<span class="kw">tolower</span>(A_clas[[<span class="dv">1</span>]]<span class="op">$</span>name) <span class="op">%in%</span><span class="st"> </span>B2]</span>
<span id="cb20-43"><a href="#cb20-43"></a><span class="co">#&gt; [1] &quot;genus&quot;</span></span>
<span id="cb20-44"><a href="#cb20-44"></a>A_clas[[<span class="dv">1</span>]]<span class="op">$</span>rank[<span class="kw">tolower</span>(A_clas[[<span class="dv">1</span>]]<span class="op">$</span>name) <span class="op">%in%</span><span class="st"> </span>B3]</span>
<span id="cb20-45"><a href="#cb20-45"></a><span class="co">#&gt; [1] &quot;family&quot;</span></span></code></pre></div>
<p>If we find a direct match (here <em>Gammarus roeseli</em>), we are lucky. But we can also match Gammaridae with <em>Gammarus roeseli</em>, but on a lower taxonomic level. A more comprehensive and realistic example (matching a trait table with an abundance table) is given in the vignette on matching.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
